### PR TITLE
NSStrings percent decoding test

### PR DESCRIPTION
**Pull request by alotorev**
Added new tests for -[NSString stringByRemovingPercentEncoding], updated existing test for percent string encoding with PASS_EQUAL, as were suggested by fredkiefer and rfm. 

Checked tests with Apple's Foundation. Two tests have failed on the current master:
Failed test:     test09.m:57 ... Percent-encoding two-byte character
Failed test:     test09.m:67 ... Percent-encoding all characters in the string

Looks like -stringByAddingPercentEncodingWithAllowedCharacters: unable to encode a string ending with two-byte characters

### CODE REVIEW

**Comment by fredkiefer** [comment]
Looks like two tests are failing. One is this:
Failed test:     test09.m:67 ... Percent-encoding all characters in the string
13304Expected '%30%31%32%33%34%35%36%37%38%39%20%41%61%42%62%43%63%44%64%45%65%46%66%47%67%48%68%49%69%4A%6A%4B%6B%4C%6C%4D%6D%4E%6E%4F%6F%50%70%51%71%52%72%53%73%54%74%55%75%56%76%57%77%58%78%59%79%5A%7A%20%5F%2D%7E%60%21%23%24%25%26%27%28%29%2A%2B%2C%2F%3A%3B%3D%3F%40%5B%5D%E2%84%A2%E2%80%A6%C2%A9%C2%AE%C2%A3%C6%92%E2%80%B0%C2%A5%C2%A7' and got (nil)

File: Tests/base/NSString/test09.m [hunk 1]
```diff
@@ -3,55 +3,83 @@
...
+  PASS_EQUAL([urlEncodedString stringByRemovingPercentEncoding], urlDecodedString, "Percent-encoded string decoding 08");
+
+
+  //check decoding of string with the encoded part in the middle
+  urlDecodedString = [urlDecodedString stringByAppendingString:asIsString];
+  urlEncodedString = [urlEncodedString stringByAppendingString:asIsString];
+  PASS_EQUAL([urlEncodedString stringByRemovingPercentEncoding], urlDecodedString, "Percent-encoded string decoding 09");
+
 
-  urlDecodedString = @"All alphabetic characters should be encoded. Symbols should not be: !@#$%^&*()_+-=";
```
Comment (fredkiefer) [inline Tests/base/NSString/test09.m:49]: Why did you remove this special test case?

**Comment by fredkiefer** [comment]
What you did looks OK, apart from the failing tests that need to get fixed. But there is still a lot of repetition in the test code. I would have preferred to have the two PASS_EQUAL tests in the helper method (Which would require an additional parameter for the test comment). That way the main code would just call that helper over and over again.

File: Tests/base/NSString/test09.m (hunk 1 above)
Comment (alotorev) [reply Tests/base/NSString/test09.m:49]: This case covered by more extensive tests above

**Comment by alotorev** [comment]
Two tests failing because of the current implementation of the method  -characterSetWithCharactersInString: it failing to decode string with two-byte characters at the end of the string.

**Comment by rfm** [comment]
> On 27 Apr 2020, at 19:17, alotorev <notifications@github.com> wrote:
> 
> 
> Two tests failing because of the current implementation of the method -characterSetWithCharactersInString: it failing to decode string with two-byte characters at the end of the string.

Actually, the problem was with -stringByAddingPercentEncodingWithAllowedCharacters: trusting the supplied characterset to tell it which characters are allowed.
Non-ASCII characters can never be allowed, so a characterset which contained non-ascii characters needs to have them ignored.
I fixed the implementation to do that.

In fact, searching the Apple documentation I found the comment that:
  Any characters in allowedCharacters outside of the 7-bit ASCII range are ignored.
I think that's what my change does.

**Comment by fredkiefer** [comment]
Looks great now!
You should merge in master to get the tests to pass and then feel free to merge your changes.

**Comment by alotorev** [comment]
> Looks great now!
> You should merge in master to get the tests to pass and then feel free to merge your changes.

I did. But I have no access rights to merge.

**Comment by rfm** [comment]
> On 28 Apr 2020, at 10:08, alotorev <notifications@github.com> wrote:
> 
> 
> Looks great now!
> You should merge in master to get the tests to pass and then feel free to merge your changes.
> 
> I did. But I have no access rights to merge.

I merged ... thanks.
